# TP NoSQL

Prénom :Daniel

Nom : DA SILVA

Pour ce TP noté vous avez trois heures pour le réaliser et 48 heures pour le finir si besoin. Le rendu final est attendu pour dimanche 29 septembre 2024 à 23 h 59.

Vous serrez évalués sur les requêtes mais également sur la qualité et la propreté du code. Pour chaque question, une réponse présentant les résultats est attendue.

In [3]:
# installation des packages pandas et pymongo
!pip install -r requirements.txt --quiet

In [5]:
import pymongo
import pandas as pd

# URI = 'MONGO CONNEXTION STRING TO REPLACE'
client = pymongo.MongoClient("mongodb+srv://user_aaa:OHQnV3zE8ATqmslU@cluster0.vrwpg.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0")
db = client.tp

# output the name of the collections in the database
print("Collections: ", db.list_collection_names())

Collections:  ['restaurants', 'movies']


1. Combien de films sont dans la base de données ?

In [16]:
print(db.movies.count_documents({}))


23530


il y a 23530 films dans la base de données 

2. Quel est le nombre moyen de votes par pays ? Trier les pays par nombre moyen de votes décroissant.

In [76]:
results = db.movies.aggregate([
    { "$unwind": "$countries" },
    { "$group": { "_id": "$countries", "avg_votes": { "$avg": "$imdb.votes" } } },
    { "$sort": { "avg_votes": -1 } }
])
pd.DataFrame(list(results))


_id      avg_votes
0                   Malta  210571.800000
1                 Bahamas  206061.000000
2             New Zealand   66589.269231
3    United Arab Emirates   53495.833333
4                     USA   35016.703663
..                    ...            ...
153                Kosovo     104.500000
154               Moldova     104.000000
155           North Korea      92.666667
156                Brunei      51.000000
157               Vanuatu       6.000000

[158 rows x 2 columns]

le nombre moyen de votes à malte est de 210571.8

3. Donner la liste des différents genres de films.

In [78]:
results = db.movies.distinct("genres")
pd.DataFrame(list(results))

0
0        Action
1     Adventure
2     Animation
3     Biography
4        Comedy
5         Crime
6   Documentary
7         Drama
8        Family
9       Fantasy
10    Film-Noir
11      History
12       Horror
13        Music
14      Musical
15      Mystery
16         News
17      Romance
18       Sci-Fi
19        Short
20        Sport
21    Talk-Show
22     Thriller
23          War
24      Western

voici les différents genres de films :'Action','Adventure','Animation','Biography','Comedy','Crime','Documentary','Drama','Family','Fantasy','Film-Noir','History','Horror','Music','Musical','Mystery','News','Romance','Sci-Fi','Short','Sport','Talk-Show','Thriller','War','Western'

4. Combien de films ont une note IMDb supérieure à 8 ?

In [27]:
results =db.movies.count_documents({
    "imdb.votes": { "$gt": 8 }
})
results

23437

23437 ont une note IMDb supérieure à 8 

5. Afficher le titre, l'année de sortie et les réalisateurs de tous les films (ne pas afficher l'ID).

In [80]:
results = db.movies.find({}, { "title": 1, "year": 1, "directors": 1, "_id": 0 })

pd.DataFrame(list(results))

title  \
0      Intolerance: Love's Struggle Throughout the Ages   
1                             The Poor Little Rich Girl   
2                                               The Kid   
3                                        Miss Lulu Bett   
4                                          Safety Last!   
...                                                 ...   
23525                                            Emelie   
23526                                    Techo y comida   
23527                                             Nahid   
23528                                  Coin Locker Girl   
23529                                 Tryapichnyy soyuz   

                            directors  year  
0                     [D.W. Griffith]  1916  
1                  [Maurice Tourneur]  1917  
2                   [Charles Chaplin]  1921  
3               [William C. de Mille]  1921  
4      [Fred C. Newmeyer, Sam Taylor]  1923  
...                               ...   ...  
23525                [Michael Thelin]  2015  
23526      [Juan Miguel del Castillo]  2015  
23527                [Ida Panahandeh]  2015  
23528                   [Jun-hee Han]  2015  
23529            [Mikhail Mestetskiy]  2015  

[23530 rows x 3 columns]

le film "Intolerance: Love's Struggle Throughout the Ages" a été réalisé par 'D.W. Griffith en 1916

6. Afficher le titre et le plot des films réalisés par D.W. Griffith.

In [70]:
bk = db.movies.find(
    { "directors": "D.W. Griffith" },
    { "_id": 0, "title": 1 , "plot": 1}
)
pd.DataFrame(list(bk))

plot  \
0  The story of a poor young woman, separated by ...   
1  A greedy tycoon decides, on a whim, to corner ...   
2  A frail waif, abused by her brutal boxer fathe...   
3  The Civil War divides friends and destroys fam...   

                                              title  
0  Intolerance: Love's Struggle Throughout the Ages  
1                                 A Corner in Wheat  
2    Broken Blossoms or The Yellow Man and the Girl  
3                             The Birth of a Nation

D.W. Griffith a rélisé le film A Corner in Wheat avec le début de plot : A greedy tycoon decides, on a whim, to corner ...

7. Quels sont les 5 acteurs les plus populaires (ceux qui ont joué dans le plus de films) ?

In [100]:
top5 = db.movies.aggregate([
    { "$unwind": "$cast" },
    { "$group": { "_id": "$cast", "nb": { "$sum": 1 }}},
    { "$sort": { "nb": -1 }},
    { "$limit": 5}
])
pd.DataFrame(list(top5))

_id  nb
0      Gèrard Depardieu  68
1        Robert De Niro  60
2         Michael Caine  52
3  Marcello Mastroianni  50
4          Bruce Willis  49

Gèrard Depardieu a joué dans 68 films

8. Quelles sont les 6 années avec le plus de films produits ?

In [86]:
top5=db.movies.aggregate([
    { "$group": {"_id": "$year", "nb": { "$sum": 1 }}}, 
    { "$sort": { "nb": -1}}, 
    { "$limit": 6 }  
])

pd.DataFrame(list(top5))

_id    nb
0  2013  1220
1  2014  1147
2  2012  1109
3  2011  1040
4  2009   990
5  2010   970

en 2013 il y a eu 1220 films

9. Combien de films ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7 ?

In [93]:
top5 = db.movies.count_documents({
    "year": { "$lt": 2000 },
    "imdb.rating": { "$gt": 7 }
})
top5

4747

il y a 4747  films qui ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7

10. Donner les 20 réalisateurs qui ont réalisé le plus de films (afficher seulement leur nom et le nombre de films).

In [102]:
top5 = db.movies.aggregate([
    { "$unwind": "$directors" },
    { "$group": { "_id": "$directors", "nb": { "$sum": 1 }}},
    { "$sort": { "nb": -1 }},
    { "$limit": 20}
])
pd.DataFrame(list(top5))

_id  nb
0         Woody Allen  40
1           John Ford  35
2       Takashi Miike  34
3         John Huston  34
4       Werner Herzog  33
5     Martin Scorsese  32
6    Alfred Hitchcock  31
7        Sidney Lumet  30
8       Michael Apted  29
9    Steven Spielberg  29
10       George Cukor  29
11    Mario Monicelli  29
12      Robert Altman  28
13          Spike Lee  28
14  Steven Soderbergh  28
15          Ken Loach  27
16     Clint Eastwood  27
17        Wim Wenders  27
18         Johnnie To  27
19    Jean-Luc Godard  27

Woody Allen a réalisé 40 films

11. Quelle est la note IMDb moyenne des films sortis après 2010 ?

In [129]:
borough_nb_rest_avg_score = db.movies.aggregate([
    {"$match" :{"year": { "$gt": 2010 }}},
    { "$group": {
        "_id":  "moyenne",
        "score": { "$avg": "$imdb.rating" }
    }},
])
pd.DataFrame(list(borough_nb_rest_avg_score)).round(2)

_id  score
0  moyenne   6.54

la moyenne des films sortis après 2010 est de 6.54

12. (bonus) Afficher les films dont le titre contient le mot "love".

In [193]:
love= db.movies.find (
    { "title" : {"$regex" : "^love" , "$options" : "si"}},
     {"_id": 0,  "title": 1}
)

pd.DataFrame(list(love))

title
0              Love Finds Andy Hardy
1                    Love Me Tonight
2                   Love on the Dole
3    Love Is a Many-Splendored Thing
4                Love Me or Leave Me
..                               ...
123              Love at First Fight
124                      Love Steaks
125        Love Is the Perfect Crime
126                        Lovemilla
127                             Love

[128 rows x 1 columns]

On a 128 avec le mot "love"

13. (bonus) Afficher le titre des films ayant un score Rotten Tomatoes (viewer) supérieur à 75.


In [210]:
borough_nb_rest_avg_score = db.movies.aggregate([
    { "$unwind": "$title" },
    {"$match" :{"tomatoes.viewer.meter": { "$gt": 75 }}},
    { "$project": { "title": 1, "tomatoes.viewer.meter":1 }}

])
df=pd.DataFrame(list(borough_nb_rest_avg_score))
df = df.drop('_id', axis=1)
df

title  \
0     Intolerance: Love's Struggle Throughout the Ages   
1                            The Poor Little Rich Girl   
2                                              The Kid   
3                                         Safety Last!   
4                                         Wild Oranges   
...                                                ...   
6597                                             Luton   
6598                                       Amira & Sam   
6599                                     Misconception   
6600                                             Grace   
6601                         Batman: Assault on Arkham   

                        tomatoes  
0      {'viewer': {'meter': 78}}  
1      {'viewer': {'meter': 77}}  
2      {'viewer': {'meter': 96}}  
3      {'viewer': {'meter': 93}}  
4      {'viewer': {'meter': 91}}  
...                          ...  
6597  {'viewer': {'meter': 100}}  
6598   {'viewer': {'meter': 76}}  
6599  {'viewer': {'meter': 100}}  
6600   {'viewer': {'meter': 86}}  
6601   {'viewer': {'meter': 86}}  

[6602 rows x 2 columns]

The Poor Little Rich Girl	a une note de 77 sur Rotten Tomatoes (viewer)